aperturedb/Subscriptable.py

```diff
@@ -4,13 +4,14 @@ class Subscriptable():
     """
 
     def __getitem__(self, subscript):
+        batch_context = {}
         if isinstance(subscript, slice):
             start = subscript.start if subscript.start else 0
             stop = subscript.stop if subscript.stop else len(self)
             step = subscript.step if subscript.step else 1
-            return [self.getitem(i) for i in range(start, stop, step)]
+            return [self.getitem(i,batch_context) for i in range(start, stop, step)]
         else:
-            return self.getitem(subscript)
+            return self.getitem(subscript,batch_context)
 
-    def getitem(self, subscript):
+    def getitem(self, subscript, batch_context):
```

Comment:
This provides a mechanism for subclasses to share information between items within the same batch, so, for example, multiple commands can use the same references. The change allows for significantly more efficient batching, but it is "unpythonic" in that it can affect the behavior of `Subscriptable` subclasses in unexpected ways. I think it warrants further discussion on the merits/costs of this change, and alternative implementations.

### CON:

it can create subtle unexpected differences in behavior where, for example:
```py
bbox_data_csv[0:1000]
```
and
```py
[bbox_data_csv[i] for i in range(0:1000)]
```
produce different results.

### PRO:

It enables reference caching, which significantly speeds up data ingestion. The difference was almost 10x for COCO bboxes on my dev box:

- without batch_context, one `FindImage` for each `AddBoundingBox`:
```
- Loading bounding boxes
 Progress: 100.00% - ETA(s): 0.00 
 ============ ApertureDB Loader Stats ============
Total time (s): 556.807008266449
Total queries executed: 368
Avg Query time (s): 11.3530900258085
Query time std: 3.5559611995361515
Avg Query Throughput (q/s)): 0.7046539736595004
Overall insertion throughput (element/s): 66.0569990211028
Total inserted elements: 36781
=================================================
```

-  with this change, many `AddBoundingBox` cmds share the same `FindImage` ref:
```
- Loading bounding boxes
 Progress: 100.00% - ETA(s): 2.39 
 ============ ApertureDB Loader Stats ============
Total time (s): 63.846665143966675
Total queries executed: 368
Avg Query time (s): 1.152759766449099
Query time std: 0.8764800250987376
Avg Query Throughput (q/s)): 6.939867466612565
Overall insertion throughput (element/s): 576.0833383711303
Total inserted elements: 36781
=================================================
```